### PR TITLE
Use FMA for CUDA Adam kernel's lerps.

### PR DESCRIPTION
**Pull request by MeetThePatel**
Switch the calculation of lerps in Adam's fused CUDA kernel to use fma, as proposed by @crcrpar .

### CODE REVIEW

**Review by crcrpar — COMMENTED** [review]
Would you have perf numbers of before/after?

**Review by ngimel — DISMISSED** [review]

**Comment by MeetThePatel** [comment]
@Skylion007 I'm not sure if this is quite exactly what you're asking, but on Godbolt NVCC 12.5.1, with the example code:
```c++
double testing_fma(double a, double b, double c) {
    return std::fma(a, b, c);
}

double testing_manual(double a, double b, double c) {
    return a * b + c;
}
```

the output for `-O3` is:
```asm
testing_fma(double, double, double):
        jmp     fma
testing_manual(double, double, double):
        mulsd   %xmm1, %xmm0
        addsd   %xmm2, %xmm0
        ret
```

and the output for `-O0` is:
```asm
testing_fma(double, double, double):
        pushq   %rbp
        movq    %rsp, %rbp
        subq    $32, %rsp
        movsd   %xmm0, -8(%rbp)
        movsd   %xmm1, -16(%rbp)
        movsd   %xmm2, -24(%rbp)
        movsd   -24(%rbp), %xmm1
        movsd   -16(%rbp), %xmm0
        movq    -8(%rbp), %rax
        movapd  %xmm1, %xmm2
        movapd  %xmm0, %xmm1
        movq    %rax, %xmm0
        call    fma
        movq    %xmm0, %rax
        movq    %rax, %xmm0
        leave
        ret
testing_manual(double, double, double):
        pushq   %rbp
        movq    %rsp, %rbp
        movsd   %xmm0, -8(%rbp)
        movsd   %xmm1, -16(%rbp)
        movsd   %xmm2, -24(%rbp)
        movsd   -8(%rbp), %xmm0
        mulsd   -16(%rbp), %xmm0
        addsd   -24(%rbp), %xmm0
        movq    %xmm0, %rax
        movq    %rax, %xmm0
        popq    %rbp
        ret
```

**Comment by ngimel** [comment]
https://godbolt.org/z/oedrhM1v1 for device code (the above is host code), maps to fma.rn.f32

**Review by eqy — DISMISSED** [review]
probably wont have any unintended numerics consequences right

**Comment by ngimel** [comment]
Shouldn't. Neither original formula nor the new one recovers the value when grad and moment are equal, but we've been living with it for a long time, so should be fine

**Review by Skylion007 — DISMISSED** [review]

**Comment by ngimel** [comment]
Windows error is real `fma<double, float, float, (int)0> ` and also we shouldn't be calling fma with these mixed args? @MeetThePatel can you please take a look?

**Comment by MeetThePatel** [comment]
For the three fma's:
- `std::fma(-beta1, grad, grad)`
- `std::fma(-beta2, grad * grad, grad * grad)`
- `std::fma(beta2, exp_avg_sq, std::fma(-beta2, grad * grad, grad * grad))`

do you think it makes more sense static_cast `beta1` and `beta2` down to `opmath_t`, rather than upcasting `grad` and `exp_avg_sq` to `double`?

How much would the loss of precision on `beta1` and `beta2` be (for going from `double` to `opmath_t`).

**Comment by ngimel** [comment]
Casting betas to opmath is acceptable

**Review by Skylion007 — DISMISSED** [review]

**Comment by Skylion007** [comment]
SGD Fused Kernels could benefit from this too

**Review by janeyx99 — DISMISSED** [review]
would be great if you could run some benchmarks before and after to show any perf improvements!

**Comment by MeetThePatel** [comment]
@janeyx99 I wrote a quick benchmark: https://gist.github.com/MeetThePatel/8ad3f3591562dd781c7fb0f7e5297736

Seems to give 0.56% speedup.

**Comment by MeetThePatel** [comment]
@janeyx99 I've written a benchmark, along with results in the comments: [Benchmark Gist](https://gist.github.com/MeetThePatel/8ad3f3591562dd781c7fb0f7e5297736)

It seems that on average, this branch has a speedup of 24.87%. (note: this seems like quite a large speedup for such a trivial change; did I mess something up in the benchmark?)

**Comment by cyyever** [comment]
@MeetThePatel I thinks `std::fma` is not used, but CUDA's implementation of STL functions. If that is real, the title should be changed.

**Review by cyyever — DISMISSED** [review]

**Comment by ngimel** [comment]
Part of speedup is coming from double -> float change

**Comment by ngimel** [comment]
Can you benchmark just the optimizer, using triton.do_bench or torch.utils.benchmark?

**Comment by MeetThePatel** [comment]
@ngimel Updated the benchmark in the gist, with results in the comments.

~6.72ms -> ~6.15ms

**Comment by ngimel** [comment]
Yeah that looks reasonable, I'm still surprised at your previous results

**Comment by MeetThePatel** [comment]
> Part of speedup is coming from double -> float change

This was the case. After applying FMA to https://github.com/pytorch/pytorch/pull/154069, I'm measuring only a 2-4µs speedup from using FMA on Adam, and 0-1µs speedup on SGD.

**Comment by janeyx99** [comment]
ah right there were test failures

**Comment by MeetThePatel** [comment]
The test failures are most likely coming from the casting of the betas from `double` -> `opmath_t`. Should I remove this, and instead have the operations with the betas be done in `double`?

**Comment by ngimel** [comment]
I think we should just increase tolerances
